esp/logging/datafieldmap.cpp

```diff
@@ -0,0 +1,101 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2016 HPCC Systems.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "datafieldmap.hpp"
+const char* const defaultLogSourcePath = "Source";
+
+void ensureInputString(const char* input, bool lowerCase, StringBuffer& inputStr, int code, const char* msg)
+{
+    inputStr.set(input).trim();
+    if (inputStr.isEmpty())
+        throw MakeStringException(code, "%s", msg);
+    if (lowerCase)
+        inputStr.toLowerCase();
+}
+
+void readLogGroupCfg(IPropertyTree* cfg, StringAttr& defaultLogGroup, MapStringToMyClass<CLogGroup>& logGroups)
+{
+    StringBuffer groupName;
+    Owned<IPropertyTreeIterator> iter = cfg->getElements("LogGroup");
+    ForEach(*iter)
+    {
+        ensureInputString(iter->query().queryProp("@name"), true, groupName, -1, "LogGroup @name required");
+        Owned<CLogGroup> logGroup = new CLogGroup(groupName.str());
+        logGroup->loadMappings(iter->query());
+        logGroups.setValue(groupName.str(), logGroup);
+        if (defaultLogGroup.isEmpty())
+            defaultLogGroup.set(groupName.str());
+    }
+}
+
+void readLogSourceCfg(IPropertyTree* cfg, unsigned& logSourceCount, StringAttr& logSourcePath, MapStringToMyClass<CLogSource>& logSources)
+{
+    logSourceCount = 0;
+    Owned<IPropertyTreeIterator> iter = cfg->getElements("LogSourceMap/LogSource");
+    ForEach(*iter)
+    {
+        StringBuffer name, groupName, dbName;
```

Comment:
move these declarations outside the ForEach, re-use rather than creating new ones every iteration